templates/components/multiagent/python/app/workflows/tools.py

```diff
@@ -0,0 +1,189 @@
+            "tool_call_id": tool_call.tool_id,
+            "name": tool.metadata.get_name(),
+        }
+        if not tool:
+            tool_msgs.append(
+                ChatMessage(
+                    role=MessageRole.ASSISTANT,
+                    content=f"Tool {tool_call.tool_name} does not exist",
+                    additional_kwargs=additional_kwargs,
+                )
+            )
+            continue
+        try:
+            if show_progress and emit_agent_events:
+                ctx.write_event_to_stream(
+                    AgentRunEvent(
+                        name=agent_name,
+                        msg=f"Calling tool {tool_call.tool_name}, {tool_call.tool_kwargs}",
+                        event_type=AgentRunEventType.PROGRESS,
+                        data={
+                            "id": progress_id,
+                            "total": total_steps,
+                            "current": i,
+                        },
+                    )
+                )
+            else:
+                ctx.write_event_to_stream(
+                    AgentRunEvent(
+                        name=agent_name,
+                        msg=f"Calling tool {tool_call.tool_name}, {str(tool_call.tool_kwargs)}",
+                    )
+                )
+            if isinstance(tool, ContextAwareTool):
+                if ctx is None:
+                    raise ValueError("Context is required for context aware tool")
+                # inject context for calling an context aware tool
+                response = await tool.acall(ctx=ctx, **tool_call.tool_kwargs)
+            else:
+                response = await tool.acall(**tool_call.tool_kwargs)  # type: ignore
+            tool_msgs.append(
+                ChatMessage(
+                    role=MessageRole.TOOL,
+                    content=str(response.raw_output),
+                    additional_kwargs=additional_kwargs,
+                )
+            )
+        except Exception as e:
+            # Print trace back here
+            tool_msg = ChatMessage(
+                role=MessageRole.TOOL,
+                content=f"Error: {str(e)}",
+                additional_kwargs=additional_kwargs,
+            )
+            ctx.write_event_to_stream(
+                AgentRunEvent(
+                    name=agent_name,
+                    msg=f"Error in tool {tool_call.tool_name}: {str(e)}",
+                )
+            )
+            tool_msgs.append(tool_msg)
+    if show_progress and emit_agent_events:
+        ctx.write_event_to_stream(
+            AgentRunEvent(
+                name=agent_name,
+                msg="Task finished",
+            )
+        )
+
+    return tool_msgs
+
+
+async def tool_calls_or_response(  # type: ignore
+    llm: FunctionCallingLLM,
+    tools: list[BaseTool],
+    chat_history: list[ChatMessage],
+) -> tuple[
+    list[ToolSelection] | None, AsyncGenerator[ChatMessage, None] | ChatResponse
+]:
```

Comment:
that's already cleaner:
```suggestion
class ToolCalls:
  calls: list[ToolSelection]
  msg: ChatMessage
  
async def chat_with_tools(  # type: ignore
    llm: FunctionCallingLLM,
    tools: list[BaseTool],
    chat_history: list[ChatMessage],
) -> ToolCalls | AsyncGenerator[ChatMessage, None]:
```

Reply:
> ToolCalls | AsyncGenerator[ChatMessage, None]:

I think it doesn't work with Python; a function must either return a result or be a generator.
I updated to this:

```python
class ResponseGenerator(BaseModel):
    """
    A response generator from chat_with_tools.
    """

    generator: AsyncGenerator[ChatResponse | None, None]

    model_config = ConfigDict(arbitrary_types_allowed=True)


class ToolCallResponse(BaseModel):
    """
    A tool call response from chat_with_tools.
    """

    tool_calls: list[ToolSelection]
    tool_call_message: ChatMessage
 
 ...
 
async defchat_with_tools() ->  ToolCallResponse | ResponseGenerator
    
```